src/test/java/com/google/cloud/spanner/pgadapter/JdbcSimpleModeMockServerTest.java

```diff
@@ -172,83 +177,327 @@ public void testQueryWithParameters() throws SQLException {
+        mockSpanner.getRequestsOfType(ExecuteBatchDmlRequest.class);
+    assertEquals(1, requests.size());
+    ExecuteBatchDmlRequest request = requests.get(0);
+    assertEquals(2, request.getStatementsCount());
+    assertEquals(INSERT_STATEMENT.getSql(), request.getStatements(0).getSql());
+    assertEquals(INVALID_DML.getSql(), request.getStatements(1).getSql());
+
+    // Verify that the implicit transaction is rolled back due to the exception
+    List<RollbackRequest> rollbackRequests = mockSpanner.getRequestsOfType(RollbackRequest.class);
+    assertEquals(1, rollbackRequests.size());
+  }
+
+  @Test
+  public void testErrorHandlingInExplicitTransactionWithCommit() throws SQLException {
     String sql =
-        "CREATE TABLE testtable (id integer PRIMARY KEY, data integer); INSERT INTO testtable (id, data) VALUES (1, 2); INSERT INTO testtable (id, data) VALUES (3, 4);";
+        String.format(
+            "%s; %s; %s; %s; %s;",
+            INSERT_STATEMENT, "BEGIN", UPDATE_STATEMENT, INVALID_DML, "COMMIT");
     try (Connection connection = DriverManager.getConnection(createUrl())) {
-      try (Statement statement = connection.createStatement()) {
+      try (java.sql.Statement statement = connection.createStatement()) {
         SQLException exception = assertThrows(SQLException.class, () -> statement.execute(sql));
-        assertEquals(
-            String.format(
-                "ERROR: FAILED_PRECONDITION: Executing updates is not allowed for DDL batches. \"%s\"",
-                sql.substring(0, sql.length() - 1)),
-            exception.getMessage());
+        assertThat(
+            exception.getMessage(), containsString("INVALID_ARGUMENT: Statement is invalid."));
+
+        // Verify that the explicit transaction is aborted due to the exception
+        SQLException exception2 =
+            assertThrows(SQLException.class, () -> statement.execute(SELECT1.getSql()));
+        assertThat(
+            exception2.getMessage(),
+            containsString(
+                "Transaction is aborted and must be rolled back prior to further execution."));
       }
     }
+
+    // Verify that the DML statements were batched together by PgAdapter.
+    List<ExecuteSqlRequest> requests = mockSpanner.getRequestsOfType(ExecuteSqlRequest.class);
+    assertEquals(1, requests.size());
+    assertEquals(INSERT_STATEMENT.getSql(), requests.get(0).getSql());
+
+    List<ExecuteBatchDmlRequest> batchDmlRequests =
+        mockSpanner.getRequestsOfType(ExecuteBatchDmlRequest.class);
+    assertEquals(1, requests.size());
+    ExecuteBatchDmlRequest request = batchDmlRequests.get(0);
+    assertEquals(2, request.getStatementsCount());
+    assertEquals(UPDATE_STATEMENT.getSql(), request.getStatements(0).getSql());
+    assertEquals(INVALID_DML.getSql(), request.getStatements(1).getSql());
+
+    List<RollbackRequest> rollbackRequests = mockSpanner.getRequestsOfType(RollbackRequest.class);
+    assertEquals(0, rollbackRequests.size());
```

Comment:
There should be 1 rollback request. The `COMMIT` statement in the batch should not be ignored by PGAdapter, but should be changed into a `ROLLBACK`.

Reply:
Done.